File: c++/Binary Tree Postorder Traversal.cpp
#include <iostream>
#include <vector>
#include <stack>

struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution {
public:
    std::vector<int> postorderTraversal(TreeNode* root) {
        std::vector<int> result;
        if (root == nullptr) return result;

        std::stack<TreeNode*> nodes;
        nodes.push(root);

        while (!nodes.empty()) {
            TreeNode* curr = nodes.top();
            nodes.pop();
            result.insert(result.begin(), curr->val);

            if (curr->left != nullptr) {
                nodes.push(curr->left);
            }
            if (curr->right != nullptr) {
                nodes.push(curr->right);
            }
        }

        return result;
    }
};
// test case
int main() {
    Solution solution;
    TreeNode* root = new TreeNode(1);
    root->right = new TreeNode(2);
    root->right->left = new TreeNode(3);

    std::vector<int> postorder = solution.postorderTraversal(root);

    std::cout << "postorder traversal: ";
    for (int val : postorder) {
        std::cout << val << " ";
    }
    std::cout << std::endl;

    delete root->right->left;
    delete root->right;
    delete root;

    return 0;
}

File: python/Binary Tree Postorder Traversal.py
from typing import List, Optional

class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution:
    def postorderTraversal(self, root: Optional[TreeNode]) -> List[int]:
        result = []
        if root is None:
            return result

        nodes = [root]

        while nodes:
            curr = nodes.pop()
            result.insert(0, curr.val)

            if curr.left is not None:
                nodes.append(curr.left)
            if curr.right is not None:
                nodes.append(curr.right)

        return result
    
# Test cases
def run_tests():
    solution = Solution()

    # case 1
    root1 = None
    result1 = solution.postorderTraversal(root1)
    print("Test 1 - Empty Tree:", result1)  # Expected output: []

    # case 2
    root2 = TreeNode(1)
    result2 = solution.postorderTraversal(root2)
    print("Test 2 - Single Node Tree:", result2)  # Expected output: [1]

    # case 3
    root3 = TreeNode(1, TreeNode(2), TreeNode(3))
    result3 = solution.postorderTraversal(root3)
    print("Test 3 - Simple Binary Tree:", result3)  # Expected output: [2, 3, 1]

run_tests()